webhooks: Fix repeating description for assignee change.

cozyrohan opened the pull request:
https://github.com/zulip/zulip/issues/16345

The changes were tested on my local development environment.
Two fixtures and their corresponding tests were added for
the bug fix.

cozyrohan commented:
@Hypro999 
Can we close the old pull request (#16630 ), I am creating #16678 to replace it.
I learned a lot about git, and created a branch for the issue in my local repo. 
I tried to cover all your feedback in this PR.
Thank you :)

Hypro999 commented:
We could, but in hindsight you could have just pushed back on that branch to update the PR. So something like:
```
git checkout old_branch
git reset new_branch --hard
git push origin old_branch --force
````
would have let you use a new branch but keep the same remote branch for the sake of the PR.

Hypro999 reviewed (changes requested):
Ok @cozyrohan, this PR looks a lot better now. The fixtures look more realistic and the code is a whole lot cleaner so good job! The only change I'd like to see before we can merge this is renaming the file `merge_request_hook__merge_request_updated_with_change_section_that_is_not_description.json` to `merge_request_hook__merge_request_updated_with_assignee_changes.json`. You can (and should) leave the name of the corresponding test the same though.

Edit:
In your commit message, you can drop the
> The changes were tested on my local development environment.
Two fixtures and their corresponding tests were added for
the bug fix.

part.

To fix the CI failures, run `./tools/lint --groups=backend` from your dev environment and fix the stuff it tells you (as far as I can see it's just small whitespace stuff).

cozyrohan commented:
@Hypro999 : I have resolved all the linter errors relevant to my code. I do not understand the one error that is failing the build. It does not seem to have anything to do with the code I added. Could you please guide me here.

Nov 03 08:06:04 To learn more about the test itself, see zerver/openapi/test_curl_examples.py.
Nov 03 08:06:04 
Nov 03 08:06:04 Traceback (most recent call last):
Nov 03 08:06:04   File "./tools/test-api", line 85, in <module>
Nov 03 08:06:04     test_generated_curl_examples_for_success(client)
Nov 03 08:06:04   File "/__w/zulip/zulip/zerver/openapi/test_curl_examples.py", line 68, in test_generated_curl_examples_for_success
Nov 03 08:06:04     assert(response["result"] == "success")
Nov 03 08:06:04 AssertionError
Error: Process completed with exit code 1.

Hypro999 commented:
@cozyrohan, this doesn't seem to be an issue with your code, so rest easy. I can see the same issue in [other PRs](https://github.com/zulip/zulip/pull/16685/checks?check_run_id=1347458861) too. It also looks like there's a [discussion](https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/test.20works.20locally.20but.20failing.20on.20Circle.20CI) going on about this over at CZO (chat.zulip.org).

timabbott commented:
I think that's fixed in master; just rebase and focus on issues related to your changes.  Sorry for the trouble!

cozyrohan commented:
@Hypro999 : I squashed my pull request commit history to make it cleaner. I believe I have now completed everything. Could you kindly do a final review, and let me know if there is anything more I need to address for a successful merge of my pull request

Hypro999 commented:
@cozyrohan: Well, it looks like you haven't addressed anything from my [last round of review](https://github.com/zulip/zulip/pull/16678#pullrequestreview-521381587) so could you make those changes first?

cozyrohan commented:
Sorry I missed those revisions!
 I changed the name of the test fixture from merge_request_hook__merge_request_updated_with_change_section_that_is_not_description.json  to 
merge_request_hook__merge_request_updated_with_assignee_changes.json

I also cleaned up the commit message by removing the paragraph about testing, and got all the checks to pass  :)
Is everything good to go?

Hypro999 commented:
@cozyrohan, just one last small suggestion. Can we rename the `merge_request_hook__merge_request_updated.json` to `merge_request_hook__merge_request_updated_with_missing_changes_key.json` since that makes it more clear why that fixture exists w.r.t. the new ones you've added?

Other than that everything looks good here. Once you've made that change, we can get one of the project maintainers involved and get this merged.

cozyrohan commented:
@Hypro999
I have made the last code review suggestion. 
I was going to look into this same description repeat issue for GitHub as well, should I open a new issue? Or work on it in this PR?

Hypro999 commented:
@cozyrohan great work on finishing this! I think that we should merge this PR now and close the issue as well. After you finish auditing the other git related integrations you can open a new PR for them, there's no need to open a new issue though. In the PR message you can include a reference to this PR/issue and that should be enough.

@timabbott can we get this merged now?

cozyrohan commented:
Thank you! And thank you for your guidance :)

cozyrohan commented:
@Hypro999 
Just a progress update.
I recreated the issue, inspected the json payload from the webhook, and have identified the methods that need modification. I am now working to add the code to solve the issue  : )

cozyrohan commented:
@Hypro999 - The Github test fixtures assume a standard json, that has constants defined for the Json( see below..). 

If I  had generate my own Json for the Github test fixtures, it breaks the existing tests. 

Based on how the tests are currently written would it be okay if I use the existing json as a template and add my modifications, for the issue I am trying to fix?

Current Code snippet showing tests dependence on existing Json test fixture template:
TOPIC_REPO = "public-repo"
TOPIC_ISSUE = "public-repo / Issue #2 Spelling error in the README file"
TOPIC_PR = "public-repo / PR #1 Update the README with new information"
TOPIC_DEPLOYMENT = "public-repo / Deployment on production"
TOPIC_ORGANIZATION = "baxterandthehackers organization"
TOPIC_BRANCH = "public-repo / changes"
TOPIC_WIKI = "public-repo / Wiki Pages"

class GitHubWebhookTest(WebhookTestCase):

Hypro999 commented:
In this case it's definitely alright to modify generated fixtures to maintain consistency - but I'm not really fond of the idea of just hand-tweaking an existing fixture. So what I would suggest is generate any new fixtures you need then use a systematic search-and-replace to use the values from the constants. This way we can replicate events that are actually possible with the GitHub webhook while still keeping the same values used in the other fixtures (i.e. genuine fixture structure, modified contents/strings).

Hypro999 commented:
@timabbott: just a gentle reminder that we can merge this now; perhaps after giving it a quick look though.

Hypro999 commented:
@cozyrohan: I actually have a question, how does generating new fixtures break the existing tests? You're not overwriting any of the existing fixtures are you?

cozyrohan commented:
@Hypro999 My bad, the existing tests are actually not affected at all. It is just the new tests that I am writing. 

GitHub shows me the JSON payload right in the WebHook section, so I was just taking that genuine payload directly and making the fixture like I did for GitLab. I will try doing the search and replace,  and use hand-tweaking the existing feature as a last resort. Thanks :)
